mw_url_rewrite.go

```diff
@@ -479,13 +490,39 @@ func checkPayload(r *http.Request, options apidef.StringRegexMap, triggernum int
 	contextData := ctxGetData(r)
 	bodyBytes, _ := ioutil.ReadAll(r.Body)
 
-	b := options.Check(string(bodyBytes))
-	if len(b) > 0 {
-		kn := "trigger-" + strconv.Itoa(triggernum) + "-payload"
+	matches := options.FindAllStringSubmatch(string(bodyBytes), -1)
 
-		contextData[kn] = string(b)
+	if len(matches) > 0 {
+		kn := buildTriggerKey(triggernum, "payload")
+		contextData[kn] = matches[0][0]
+
+		for i, match := range matches {
+			kn = buildTriggerKey(triggernum, "payload", i)
+			contextData[kn] = match[0]
+
+			addGroupsToContextData(&contextData, kn, match[1:])
+		}
 		return true
 	}
 
 	return false
 }
+
+func buildTriggerKey(num int, name string, indices ...int) string {
+	parts := []string{triggerKeyPrefix, strconv.Itoa(num), name}
+
+	if len(indices) > 0 {
+		for _, index := range indices {
+			parts = append(parts, strconv.Itoa(index))
+		}
+	}
+
+	return strings.Join(parts, triggerKeySep)
+}
+
+func addGroupsToContextData(cd *map[string]interface{}, keyPrefix string, groups []string) {
```

Comment:
you don't need to pass map as pointer as maps (and slices) are passed by reference

Reply:
Hmm, for me this one looks suspicious.
Intent of `addGroupsToContextData` is to modify (add items to) passed map.
As far as i know, in golang everything is value and passed by value (copied). Slices can look like passed by reference, because internally they holds pointer to baking array, first item index, length and capacity.
It can feels like they passed by reference (in case of modifying existing values), but in case of adding items they not. What's why `append()` returns updated slice.
https://play.golang.org/p/arTxwX1w9mf

I've checked same approach with map and it works, probably because in essence map is just a pointer to underlying structure - i don't know.

Will remove pointers to map as requested

Reply:
this is tricky one - slices are different because they have `len` and `cap` in addition to pointer to underlaying array with elements, so when we pass it to function it receives new "descriptor" of slice but original pointer to elements.

if we do append inside function we have to return slice and re-assign result on calling side, otherwise calling part will hold slice-descriptor with appended elements but old `len` and `cap` (thats why Println outputs only one element). you can find more info about that situation here https://golang.org/doc/effective_go.html#slices

Changing elements would work if your slice is pre-allocated, i.e.:
https://play.golang.org/p/ZoK7WLz_Xtw

That's actually how all functions like `func Read(slice []interface{}) (int, error)` reading data into passed slice without any explicit pointers.

But my comment was about maps, they work differently.